balance_strategy.go

```diff
@@ -127,3 +155,904 @@ func balanceStrategyHashValue(vv ...string) uint32 {
 	}
 	return h
 }
+
+type stickyBalanceStrategy struct {
+	movements partitionMovements
+}
+
+// Name implements BalanceStrategy.
+func (s *stickyBalanceStrategy) Name() string { return StickyBalanceStrategyName }
+
+// Plan implements BalanceStrategy.
+func (s *stickyBalanceStrategy) Plan(members map[string]ConsumerGroupMemberMetadata, topics map[string][]int32) (BalanceStrategyPlan, error) {
+	// track partition movements during generation of the partition assignment plan
+	s.movements = partitionMovements{
+		Movements:                 make(map[topicPartitionAssignment]consumerPair),
+		PartitionMovementsByTopic: make(map[string]map[consumerPair]map[topicPartitionAssignment]bool),
+	}
+
+	// prepopulate the current assignment state from userdata on the consumer group members
+	currentAssignment, prevAssignment, err := prepopulateCurrentAssignments(members)
+	if err != nil {
+		return nil, err
+	}
+
+	// determine if we're dealing with a completely fresh assignment, or if there's existing assignment state
+	isFreshAssignment := false
+	if len(currentAssignment) == 0 {
+		isFreshAssignment = true
+	}
+
+	// create a mapping of all current topic partitions and the consumers that can be assigned to them
+	partition2AllPotentialConsumers := make(map[topicPartitionAssignment][]string)
+	for topic, partitions := range topics {
+		for _, partition := range partitions {
+			partition2AllPotentialConsumers[topicPartitionAssignment{Topic: topic, Partition: partition}] = []string{}
+		}
+	}
+
+	// create a mapping of all consumers to all potential topic partitions that can be assigned to them
+	// also, populate the mapping of partitions to potential consumers
+	consumer2AllPotentialPartitions := make(map[string][]topicPartitionAssignment, len(members))
+	for memberID, meta := range members {
+		consumer2AllPotentialPartitions[memberID] = make([]topicPartitionAssignment, 0)
+		for _, topicSubscription := range meta.Topics {
+			// only evaluate topic subscriptions that are present in the supplied topics map
+			if _, found := topics[topicSubscription]; found {
+				for _, partition := range topics[topicSubscription] {
+					topicPartition := topicPartitionAssignment{Topic: topicSubscription, Partition: partition}
+					consumer2AllPotentialPartitions[memberID] = append(consumer2AllPotentialPartitions[memberID], topicPartition)
+					partition2AllPotentialConsumers[topicPartition] = append(partition2AllPotentialConsumers[topicPartition], memberID)
+				}
+			}
+		}
+
+		// add this consumer to currentAssignment (with an empty topic partition assignment) if it does not already exist
+		if _, exists := currentAssignment[memberID]; !exists {
+			currentAssignment[memberID] = make([]topicPartitionAssignment, 0)
+		}
+	}
+
+	// create a mapping of each partition to its current consumer, where possible
+	currentPartitionConsumer := make(map[topicPartitionAssignment]string, len(currentAssignment))
```

Comment:
I think the Java code and thus this transliteration from here all the way down to sortedCurrentSubscriptions on line 263 does more than necessary. I think this chunk can be similar to...
```go
167         // Maps all currently consumed partitions to the member that is
168         // currently consuming them.
169         currentPartitionConsumers := make(map[topicPartition]string)
170         // To build a list of unassigned partitions, we visit all partitions
171         // in the current plan and, if they still exist and the prior consumer
172         // no longer wants to consume them, we track it as unassigned.
173         // After, we add all new partitions.
174         unvisitedPartitions := partitions2AllPotentialConsumers.cloneKeys()
175         var unassignedPartitions []topicPartition
176 
177         for member, partitions := range currentPlan {
178                 var keepIdx int
179                 for _, partition := range partitions {
180                         // If this partition no longer exists at all, likely due to the
181                         // topic being deleted, we remove the partition from the member.
182                         if _, exists := partitions2AllPotentialConsumers[partition]; !exists {
183                                 continue
184                         }
185 
186                         delete(unvisitedPartitions, partition)
187                         currentPartitionConsumers[partition] = member
188 
189                         thisMemberIdx := sort.Search(len(members), func(i int) bool { return members[i].ID >= member })
190                         thisMember := members[thisMemberIdx]
191                         if !strsHas(thisMember.Topics, partition.topic) {
192                                 unassignedPartitions = append(unassignedPartitions, partition)
193                                 continue
194                         }
195 
196                         partitions[keepIdx] = partition
197                         keepIdx++
198                 }
199                 currentPlan[member] = partitions
200         }
201         for unvisited := range unvisitedPartitions {
202                 unassignedPartitions = append(unassignedPartitions, unvisited)
203         }
```
Types hopefully obvious.

The main point is that Java builds `unassignedPartitions` from `sortedPartitionConsumers`, which itself is built from of all of the keys in `partitions2AllPotentialConsumers`.

Rather than a bunch of logic to remove things that were added, we can just add as we noticed they are unassigned, with unvisited being everything that nothing was consuming (similar to that change in sortPartitions).

Also, rather than adding the prior assignment to currentPartitionConsumer up front, we can add as we know the partition still exists.

Other than my comment on the conditional below, I believe all of the current code here is correct, though. I'm still just trying to make sense of the Java code. I plan to enter the `balance` function tomorrow.

Reply:
Thank you for the code suggestion. I adapted it to work with this branch without much difficulty.

Note that I added a slice of consumer partition assignments to keep since the code that you provided doesn't actually remove a partition from the `partitions` slice when it's no longer subscribed to or doesn't exist.

Reply:
Hrm I thought my method slid partitions that were kept down, so that in the end after the loop, `partitions` would consist of only partitions to keep. Sort of like what you currently do in `filterAssignedPartitions`.

I do see that I forgot to trim the resulting slice to `keepIdx`, though, so a fix in mine would be `currentPlan[member] = partitions[:keepIdx]`.